### PR TITLE
Allow filtering negative ratings from ials

### PR DESCRIPTION
**Pull request by benfred**
The implicit als model assumes that non-zeros passed in are positive ratings.
For datasets like movielens passing in 1 and 2 star reviews in the training set
causes these to be learned as positive preferences - which leads to poor results.

This adds a 'train_switch_positive' option to the ialswrapper, which when set
will remove ratings from the training set less than this value.

Applying this change to the warm-start example and the iALS model starts to perform
more competitively.

### CODE REVIEW

**Comment by Evfro** [comment]
First of all, thanks for contribution!

Now, there are two different problems with this modification.

The first one is related to thresholding the data: there is a general observation, that if you exclude low ratings the behavior of any standard matrix factorization model (based on svd, sgd or als) is likely to improve. You can verify this by simply discarding all negative ratings from the data with the following line:
`data_model = RecommenderData(data.query('rating >= 4'), 'userid', 'movieid', 'rating' seed=0)`  
Both SVD and iALS will demonstrate higher scores. This is expected. Like you said, if we aim at predicting high ratings, then matrix-based models suffer from the presence of negative ratings. **I'll implement a common interface for all factorization models so that they can switch between thresholding and using all the data**.

This, however, is not always a good option. If you take tensor based model with ratings encoded along the third dimesnion, then low ratings will actually help to improve the quality of recommendations.  If you are interested, you can read more on that in our paper [Fifty Shades of Ratings](https://arxiv.org/abs/1607.04228).

The second issue is that actually in your modification higher scores are achieved due to hyper-parameter tuning (especially rank value), not due to thresholding of training data. As I mentioned in the end of the tutorial, fair comparison requires cross-validation experiment with proper hyper-parameter tuning for all compared models. I'm planning to create a new guide on this particular topic. _Would you be interested in taking part in that_? For example, you could check whether parameter tuning of iALS in that guide (when it's ready) is  appropriate.

**Comment by benfred** [comment]
Thanks for the link to your paper! I will read this weekend =).

I'm going to close this PR for now, I agree that a better solution is probably required here and that most of the gains are from tweaking the parameters. 

However, I do think that including the low rated items will hurt the iALS model more than the SVD model because of what objective function they are optimizing. The svd model will treat missing data as being a 0 star review - which is much closer to the 1 star negative reviews than a 5 star positive review is. The ALS model will treat both the 1 star and 5 star reviews as being equally positive - just with higher confidence on the positive ratings.

For  parameter tuning, I think that both the SVD and iALS model would benefit mainly from using something higher than rank 10. A guide to tuning would be great, and I'd be happy to check it when it's ready!

**Comment by Evfro** [comment]
Thanks a lot! I'll definitely let you know when the guide is ready.
By the way, I've pushed new implementation with generic value thresholding functionality to the develop branch. If you wanna play, do  
`pip install --upgrade git+https://github.com/Evfro/polara.git@develop#egg=polara`.  
There's now a new attribute `feedback_threshold`, which you can set on any model. It takes care of various edge cases with different setups. Note, it is independent of the `switch_positive` trigger, however, generally you want them be the same. The latter only affects evaluation, while the former changes the data fed into a model.

Regarding the impact of negative ratings - I agree that standard SVD will treat ratings differently. However, one can easily apply some weighting function to transform rating values and get a different scale. This is how you can tweak SVD for higher scores. Of course, it will be not the same as iALS, however, in the end, all that happens in iALS is roughly a distortion of the ratings scale. So the performance of iALS and SVD with appropriate data transformation can be made very similar. This is actually  what I observe in some of my experiments.

**Comment by recsys19** [comment]
@benfred, hi

I've finally finished the tutorial on tuning and cross-validation. So if you're still interested - check it out at:
https://github.com/Evfro/polara/blob/develop/examples/Hyper_parameter_tuning_and_cross_validation_experiments.ipynb

If you'd like to run the code yourself, make sure to update Polara again from the develop branch
`pip install --upgrade git+https://github.com/Evfro/polara.git@develop#egg=polara`

**Comment by Evfro** [comment]
@benfred, hi

I've finally finished the tutorial on tuning and cross-validation. So if you're still interested - check it out at:
https://github.com/Evfro/polara/blob/develop/examples/Hyper_parameter_tuning_and_cross_validation_experiments.ipynb

If you'd like to run the code yourself, make sure to update Polara again from the develop branch
`pip install --upgrade git+https://github.com/Evfro/polara.git@develop#egg=polara`
